tensilelite/Tensile/Source/lib/source/ContractionSolution.cpp

```diff
@@ -1190,15 +1167,32 @@ namespace TensileLite
         if(gsu > 0)
             rv.numWorkGroups.y *= gsu;
 
-        size_t skGrid    = 0;
-        auto   tiles     = problem.getNumTiles(sizeMapping);
+        size_t skGrid = 0;
+        auto   tiles  = problem.getNumTiles(sizeMapping);
         if(sizeMapping.streamK != 0 || sizeMapping.persistentKernel != 0)
         {
             AMDGPU const* pAMDGPU = dynamic_cast<AMDGPU const*>(&hardware);
             assert(pAMDGPU != nullptr && pAMDGPU->computeUnitCount != 0);
             if(sizeMapping.streamK != 0)
             {
-                skGrid             = getSKGrid(problem, hardware, tiles);
+                if(!inputs.ws)
+                {
+                    std::cerr << "\nstreamk warning: workspace not initialized, falling back to "
+                                 "data parallel mode"
+                              << std::endl;
+                    skGrid = tiles;
+                }
+                else if(inputs.workspaceSize < requiredWorkspaceSize(problem, hardware))
```

Comment:
It feels a little strange that we're using `requiredWorkspaceSize` here to determine what to set `skgrid` to, but `requiredWorkspaceSize` uses the skgrid to compute the required workspace.

It feels a little like a circular dependency.

skgrid depends on workspacesize which also depends on skgrid.

But maybe this is okay, since this is just handling an edge case when not enough workspace is passed in.

Reply:
I understand the ambiguity. To clarify, requiredWorkspaceSize considers the expected output from getSKGrid to determine the amount of workspace needed. In this scenario, instead of blindly setting skgrid value to the getSKGrid result we consider safer options when the workspace is lacking. While the naming causes a double take, there is no actual circular dependency.